Support InstantRender empty built-in

sb98052 opened the pull request:
Resolves #2186. Embeds the value `__empty` in array and object literals for properties that are conditionally set, instead of handling such values via assignments and deletes. Raises an exception if there is a cycle in object or array creation (InstantRender does not support these).

NTillmann commented on src/utils/babelhelpers.js:
```diff
@@ -21,6 +21,7 @@ import * as t from "@babel/types";
 export const voidExpression: BabelNodeExpression = t.unaryExpression("void", t.numericLiteral(0), true);
 export const nullExpression: BabelNodeExpression = t.nullLiteral();
 export const emptyExpression: BabelNodeIdentifier = t.identifier("__empty");
+export const instantRenderEmptyBuiltIn: BabelNodeIdentifier = t.identifier("__empty");
```
We don't really need this. I think `emptyExpression` would do just fine.

NTillmann commented on test/serializer/instant-render/EmptyBuiltInProps.js:
```diff
@@ -0,0 +1,19 @@
+
+inspect = function() {
+  let result = f(false);
+  if (!global.__optimize) result.foo = {};
```
`inspect` will always run in node, where `global.__optimize` doesn't exist. So I think this makes this test useless.

NTillmann commented on test/serializer/instant-render/EmptyBuiltInArray.js:
```diff
@@ -0,0 +1,19 @@
+
+inspect = function() {
+  let result = f(false);
+  if (!global.__optimize) result[0] = {};
```
See other test comment.

NTillmann commented on test/serializer/instant-render/EmptyBuiltInProps.js:
```diff
@@ -0,0 +1,19 @@
+inspect = function() {
+  let result = f(false);
+  if (!global.__optimize) result.foo = {};
+  return JSON.stringify(result);
```
Maybe `return !("foo" in result) || result.foo === __empty`. Not sure if the __empty would get rewritten. That case, you `add at runtime:var __empty = {empty: true};`, and then check if `result.foo.empty`.

sb98052 commented:
Thanks. I've addressed the comments, and this time confirmed that when // instant render is removed, the tests fail with the appropriate errors. Also some contortions to make node expose Instant Render's __empty embedded in an array, since trailing __empty builtins do not increase the length of the array.
